src/ccl_cls.c

```diff
@@ -217,50 +292,31 @@ void ccl_angular_cls_limber(ccl_cosmology *cosmo,
         // Get integration limits
         get_k_interval(cosmo, trc1, trc2, l, &lkmin, &lkmax);
 
-        // Integrate
-        gslstatus = gsl_integration_qag(&F, lkmin, lkmax, 0,
-                                        cosmo->gsl_params.INTEGRATION_LIMBER_EPSREL,
-                                        cosmo->gsl_params.N_ITERATION,
-                                        cosmo->gsl_params.INTEGRATION_LIMBER_GAUSS_KRONROD_POINTS,
-                                        w, &result, &eresult);
-
-        // Test if a round-off error occured in the evaluation of the integral
-        // If so, try another integration function, more robust but potentially slower
-        if (gslstatus == GSL_EROUND) {
-          ccl_raise_gsl_warning(
-              gslstatus,
-              "ccl_cls.c: ccl_angular_cl_limber(): "
-              "Default GSL integration failure, attempting backup method.");
-
-          if (w_cquad == NULL) {
-            w_cquad = gsl_integration_cquad_workspace_alloc(cosmo->gsl_params.N_ITERATION);
-            if (w_cquad == NULL) {
-              local_status = CCL_ERROR_MEMORY;
-            }
-          }
-
-          if (local_status == 0) {
-            nevals = 0;
-            gslstatus = gsl_integration_cquad(
-              &F, lkmin, lkmax, 0,
-              cosmo->gsl_params.INTEGRATION_LIMBER_EPSREL,
-              w_cquad, &result, &eresult, &nevals);
-          }
-        }
-
-        if (gslstatus == GSL_SUCCESS && (*ipar.status == 0) && local_status == 0) {
+	// Integrate
+	if(integration_method == ccl_integration_quad) {
+	  integ_cls_limber_quad(cosmo, &F, lkmin, lkmax, w,
+				&result, &eresult, &local_status);
+	}
+	else if(integration_method == ccl_integration_spline) {
+	  integ_cls_limber_spline(cosmo, &ipar, lkmin, lkmax,
+				  &result, &local_status);
+	}
+	else
+	  local_status = CCL_ERROR_NOT_IMPLEMENTED;
+
+        if ((*ipar.status == 0) && (local_status == 0)) {
           cl_out[lind] = result / (l+0.5);
         }
         else {
-          ccl_raise_gsl_warning(gslstatus, "ccl_cls.c: ccl_angular_cl_limber():");
+          ccl_raise_gsl_warning(local_status, "ccl_cls.c: ccl_angular_cl_limber():");
           cl_out[lind] = NAN;
           local_status = CCL_ERROR_INTEG;
         }
       }
     }
 
-    gsl_integration_workspace_free(w);
-    gsl_integration_cquad_workspace_free(w_cquad);
+    if(integration_method == ccl_integration_quad)
+      gsl_integration_workspace_free(w);
 
     if (local_status) {
       #pragma omp atomic write
```

Comment:
I forgot about atomic writes in the previous PR.

Reply:
I've now used them in ccl_haloprofiles.c